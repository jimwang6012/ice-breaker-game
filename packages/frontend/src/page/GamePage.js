import { useContext, useEffect, useState } from "react";
import { useKeyUp } from "react-keyboard-input-hook";
import classNames from "classnames";
import { useParams } from "react-router-dom";
import { useLocation, useNavigate } from "react-router-dom";
import {
  changeDirection,
  breakIce,
  breakerMove,
  playerMove,
} from "../util/PlayerControl";
import "./gampage.css";
import socket from "../Socket";
import { Modal } from "../component/Modal";
import { AppContext } from "../AppContextProvider";
import { LeaderBoard } from "../component/LeaderBoard";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faVideoCamera } from "@fortawesome/free-solid-svg-icons";
import { convertToMS } from "../util/GameTimer";

function GamePage() {
  const navigate = useNavigate();
  const roomId = useParams().code;

  const { config, isTyping } = useContext(AppContext);
  const { state } = useLocation();
  //Uses the user's config to initialize the time limit for the game
  const [currentTime, setCurrentTime] = useState(config.roundTime);
  const [show, setShow] = useState(false);

  // use for remove the out overlaying string
  const [showOut, setShowOut] = useState(true);

  const [leaderboardList, setLeaders] = useState();
  const [winningMessage, setWinner] = useState("");

  // for pre-start game initialization
  const initGame = (game, myId) => {
    const size = calculateSize();

    setBoard(game.board);
    setIceSize(size / (game.board?.length + 3));

    const playerList = game.players;
    playerList.forEach((player) => {
      if (player.id === myId) {
        setMe(player);
      }
    });
    setPlayers(playerList);
  };

  // used for page repsonsiveness
  const calculateSize = () => {
    const initWidth = window.innerWidth * 0.65;
    const initHeight = window.innerHeight - 20;
    return initWidth > initHeight ? initHeight : initWidth;
  };
  const windowSizeChanged = () => {
    const size = calculateSize();
    setIceSize(size / (board?.length + 3));
  };
  window.addEventListener("resize", windowSizeChanged);

  useEffect(() => {
    // @ts-ignore
    if (!!state.game) initGame(state.game, socket.id);

    const onGameUpdate = (game) => initGame(game, socket.id);
    const onGameEnd = (room) => {
      room.players.sort(function (a, b) {
        return b.score - a.score;
      });

      setLeaders(room.players);
      // Set the message to display and get the winner
      if (
        room.players.filter((player) => player.isAlive && !player.isBreaker)
          .length === 0
      ) {
        setWinner("The Seal Wins !!!");
      } else {
        setWinner("Penguins Win !!!");
      }
      //Open Modal window
      setShow(true);
    };
    //Update time
    const onGameTimeChanged = (time) => {
      setCurrentTime(time);
    };

    const gameDone = () => {
      setDone(true);
    };
    socket.on("room-closed", gameDone); //Set up socket on room-closed, to close the game
    socket.on("game-update", onGameUpdate);
    socket.on("game-end", onGameEnd);
    socket.on("game-time-changed", onGameTimeChanged);

    return () => {
      socket.off("room-closed", gameDone);
      socket.off("game-update", onGameUpdate);
      socket.off("game-end", onGameEnd);
      socket.off("game-time-changed", onGameTimeChanged);
    };
    // @ts-ignore
  }, [navigate, roomId, state.game]);

  // used for game state checking
  const [players, setPlayers] = useState([]);
  const [iceSize, setIceSize] = useState(10);
  const [done, setDone] = useState(false);

  // default player parameters
  const [me, setMe] = useState({
    id: null,
    name: null,
    isAlive: true,
    isBreaker: false,
    x: 0,
    y: 0,
    direction: null,
  });
  const [board, setBoard] = useState([]);

  // handle local player moves
  const handlePlayerMove = ({ keyName }) => {
    if (me.isAlive && !done && !isTyping) {
      // what every move or not, direction need to change
      if (me.isBreaker && !isBreaking) {
        changeDirection(me, keyName);
        if (keyName === "Space") {
          //handle break ice using space
          breakIce(me, board, handleBreak);
        } else {
          // movement of breakers
          breakerMove(keyName, me, board);
        }
      } else if (!me.isBreaker) {
        changeDirection(me, keyName);
        // normal non-breaker players on the ice
        playerMove(keyName, me, board.length);
      }
      socket.emit("movement", {
        roomId,
        x: me.x,
        y: me.y,
        direction: me.direction,
      });
      setMe(me);
    }
  };

  // hook the user keyboard listener up
  useKeyUp(handlePlayerMove);

  // disable breaker move if currently is breaking a ice
  const [isBreaking, setIsBreaking] = useState(false);

  // breaking ice controls
  const handleBreak = async (row, col) => {
    setIsBreaking(true);
    await new Promise((resolve) =>
      setTimeout(resolve, config.breakTime * 1000)
    );
    board[row][col] = 0;
    setBoard([...board]);
    socket.emit("break", { roomId, y: col, x: row });
    setIsBreaking(false);
  };

  // determine whether input players on this ice by compare x and y
  const onThisIce = (player, row, col) => {
    return player.x === row && player.y === col;
  };

  return (
    <div>
      {/* top right symbol for indicate the player out, which is "spectating" */}
      {!me.isAlive && (
        <div className="absolute flex flex-col p-5 text-ice-3">
          <FontAwesomeIcon style={{ fontSize: 48 }} icon={faVideoCamera} />
          <div className="font-bold">Spectating</div>
        </div>
      )}
      <div className="flex items-center justify-center h-screen overflow-y-auto bg-water-background">
        {/* Modal window for the leaderboard when game ends */}
        <Modal
          title={
            <div className="text-3xl font-black text-ice-6">
              {winningMessage}
            </div>
          }
          show={show}
          pageJump={() => {
            navigate("/" + roomId.toString() + "/idle");
          }}
          mainPrompt={<LeaderBoard list={leaderboardList} myID={me.id} />}
          buttonPrompt={"Back to Room"}
        />
        {/*Timer for the game*/}
        <div className="text-3xl font-bold text-white timer">
          {convertToMS(currentTime)}
        </div>

        <div className="flex items-center justify-center mt-4 overflow-y-auto">
          {/* Dead indicator, overlaying string */}
          {!me.isAlive && showOut && (
            <div className="absolute z-40 flex flex-row animate-bounce">
              <div
                style={{
                  fontSize: 100,
                  textShadow:
                    "-2px 0 #041F32, 0 2px #FFFFFF, 2px 0 #041F32, 0 -2px #041F32",
                }}
                className="text-ice-5 text-9xl"
                onClick={() => setShowOut(false)}
              >
                You're out!
              </div>
            </div>
          )}
          <div className="">
            {board ? (
              board.map((rowItems, row) => {
                return (
                  <div className="flex flex-row" key={row}>
                    {rowItems.map((ice, col) => {
                      return (
                        <div className="flex" key={col}>
                          {/* {If is 0 but with less than 2 players} */}
                          {ice === 1 ? (
                            <div
                              // for page resonsiveness
                              style={{
                                width: Math.round(iceSize),
                                height: Math.round(iceSize),
                                margin: Math.round(iceSize / board.length),
                              }}
                              className="flex items-center justify-center m-2 bg-no-repeat bg-cover shadow-md player bg-ice-full"
                            >
                              {players.map((item, index) => {
                                if (onThisIce(item, row, col)) {
                                  return (
                                    <PlayerAvatar
                                      key={index}
                                      player={item}
                                      ice={iceSize}
                                    />
                                  );
                                } else return null;
                              })}
                            </div>
                          ) : (
                            <div
                              // for page resonsiveness
                              style={{
                                width: Math.round(iceSize),
                                height: Math.round(iceSize),
                                margin: Math.round(iceSize / board.length),
                              }}
                              className="flex items-center justify-center m-2 bg-no-repeat bg-cover player bg-ice-broken"
                            >
                              {players.map((item, index) => {
                                if (
                                  onThisIce(item, row, col) &&
                                  item.isBreaker
                                ) {
                                  return (
                                    <PlayerAvatar
                                      key={index}
                                      player={item}
                                      ice={iceSize}
                                    />
                                  );
                                } else return null;
                              })}
                            </div>
                          )}
                        </div>
                      );
                    })}
                  </div>
                );
              })
            ) : (
              <></>
            )}
          </div>
        </div>
      </div>
    </div>
  );
}

export default GamePage;

// Get imgae path based on players' chosen colors and direction
function getIcon(props, colors) {
  const myColor = colors[props.player.colorId ?? 0]; //Obtain each player's color from the color list
  var direction = props.player.direction;
  // Reverse the direction of LEFT and RIGHT
  if (props.player.direction === "LEFT") {
    direction = "RIGHT";
  } else if (props.player.direction === "RIGHT") {
    direction = "LEFT";
  }
  // Return the img path
  return props.player.isBreaker
    ? "../avatars/seal_" + direction.toLowerCase() + ".png"
    : "../avatars/p_" +
        direction.toLowerCase() +
        "_" +
        myColor.toString().substring(1) +
        ".png";
}

// Display user Image on the board
function PlayerAvatar(props) {
  //Get the color list for players from the context
  const { colors } = useContext(AppContext);

  const img_path = getIcon(props, colors);

  if (props.player.isAlive) {
    return (
      <div
        className={classNames(
          "playerme flex items-center justify-center text-white rounded-2xl text "
        )}
      >
        <div
          className="flex items-center justify-center "
          style={{ width: props.ice * 0.85, height: props.ice * 0.85 }}
        >
          {/* Use created asset img in public/avatars/ with responsive size */}
          <img
            src={process.env.PUBLIC_URL + img_path}
            width={props.ice * 0.8}
            alt="player"
          />
        </div>
      </div>
    );
  } else {
    return null;
  }
}
