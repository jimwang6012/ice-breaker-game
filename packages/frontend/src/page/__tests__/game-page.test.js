import { BrowserRouter } from "react-router-dom";
import GamePage from "../IdlePage";
import { render } from "@testing-library/react";

it("renders correctly", () => {
  const tree = render(
    <BrowserRouter>
      <GamePage />
    </BrowserRouter>
  );
  expect(tree).toMatchSnapshot();
});
