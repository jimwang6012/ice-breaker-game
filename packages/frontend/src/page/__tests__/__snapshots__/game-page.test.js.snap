// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class=" items-center h-screen bg-ice-8 opacity-90"
      >
        <div
          class="flex flex-col items-center justify-center w-full gap-20 pt-5 "
        >
          <div
            class="text-6xl font-bold text-white"
          >
            Wait for host to Start!
          </div>
          <div
            class="mantine-Group-root mantine-1u7f61y"
            style="border-width: 3px; border-radius: 10px; border-color: transparent;"
          >
            <div
              class="mantine-1vp280o mantine-Group-child w-2/5 py-16 rounded-lg shadow-lg px-28 bg-ice-2 "
            >
              <div
                class="flex flex-row justify-center gap-2 text-2xl font-bold text-black"
              >
                Room Code: 
                
                <button>
                  <svg
                    aria-hidden="true"
                    class="svg-inline--fa fa-clipboard hover:text-ice-7"
                    data-icon="clipboard"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    viewBox="0 0 384 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M336 64h-53.88C268.9 26.8 233.7 0 192 0S115.1 26.8 101.9 64H48C21.5 64 0 85.48 0 112v352C0 490.5 21.5 512 48 512h288c26.5 0 48-21.48 48-48v-352C384 85.48 362.5 64 336 64zM192 64c17.67 0 32 14.33 32 32c0 17.67-14.33 32-32 32S160 113.7 160 96C160 78.33 174.3 64 192 64zM272 224h-160C103.2 224 96 216.8 96 208C96 199.2 103.2 192 112 192h160C280.8 192 288 199.2 288 208S280.8 224 272 224z"
                      fill="currentColor"
                    />
                  </svg>
                </button>
              </div>
              <div
                class="mantine-1eabysl"
              />
              <div
                class="mantine-1eabysl"
              />
              <div
                class="mantine-10dh38u"
              />
              <div
                class="mantine-Group-root mantine-3suwep"
              >
                <div
                  class="mantine-Group-root mantine-1u7f61y"
                >
                  <div
                    class="mantine-SimpleGrid-root mantine-Group-child mantine-1qxm3qj"
                  />
                </div>
                <div
                  class="mantine-Group-root mantine-Group-child mantine-2v0bxn"
                >
                  <div
                    class="flex"
                  >
                    <button
                      class="absolute z-40 px-6 py-3 text-xl font-semibold text-white rounded-lg shadow-inner bg-ice-6 hover:mt-1 active:mt-2"
                    >
                      Ready!
                    </button>
                    <div
                      class="px-6 py-3 mt-2 text-xl font-semibold rounded-lg text-ice-7 bg-ice-7"
                    >
                      Ready!
                    </div>
                  </div>
                  <p
                    class="mantine-gnzaph mantine-Group-child text-center w-full text-red-500 hidden"
                  >
                    Please choose a valid color
                  </p>
                </div>
              </div>
            </div>
          </div>
          <div
            class="flex flex-col items-center justify-around w-3/5 xl:flex-row gap-y-2"
          >
            <div
              class="flex"
            >
              <button
                class="absolute z-40 px-6 py-3 text-xl font-semibold text-white rounded-lg shadow-inner bg-ice-6 hover:mt-1 active:mt-2"
              >
                Leave Room
              </button>
              <div
                class="px-6 py-3 mt-2 text-xl font-semibold rounded-lg text-ice-7 bg-ice-7"
              >
                Leave Room
              </div>
            </div>
            <div
              class="flex"
            >
              <button
                class="absolute z-40 px-6 py-3 text-xl font-semibold text-white rounded-lg shadow-inner bg-ice-6 hover:mt-1 active:mt-2"
              >
                Game Config
              </button>
              <div
                class="px-6 py-3 mt-2 text-xl font-semibold rounded-lg text-ice-7 bg-ice-7"
              >
                Game Config
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        dir="ltr"
        style="position: relative; z-index: 200;"
      />
    </div>
  </body>,
  "container": <div>
    <div
      class=" items-center h-screen bg-ice-8 opacity-90"
    >
      <div
        class="flex flex-col items-center justify-center w-full gap-20 pt-5 "
      >
        <div
          class="text-6xl font-bold text-white"
        >
          Wait for host to Start!
        </div>
        <div
          class="mantine-Group-root mantine-1u7f61y"
          style="border-width: 3px; border-radius: 10px; border-color: transparent;"
        >
          <div
            class="mantine-1vp280o mantine-Group-child w-2/5 py-16 rounded-lg shadow-lg px-28 bg-ice-2 "
          >
            <div
              class="flex flex-row justify-center gap-2 text-2xl font-bold text-black"
            >
              Room Code: 
              
              <button>
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-clipboard hover:text-ice-7"
                  data-icon="clipboard"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 384 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M336 64h-53.88C268.9 26.8 233.7 0 192 0S115.1 26.8 101.9 64H48C21.5 64 0 85.48 0 112v352C0 490.5 21.5 512 48 512h288c26.5 0 48-21.48 48-48v-352C384 85.48 362.5 64 336 64zM192 64c17.67 0 32 14.33 32 32c0 17.67-14.33 32-32 32S160 113.7 160 96C160 78.33 174.3 64 192 64zM272 224h-160C103.2 224 96 216.8 96 208C96 199.2 103.2 192 112 192h160C280.8 192 288 199.2 288 208S280.8 224 272 224z"
                    fill="currentColor"
                  />
                </svg>
              </button>
            </div>
            <div
              class="mantine-1eabysl"
            />
            <div
              class="mantine-1eabysl"
            />
            <div
              class="mantine-10dh38u"
            />
            <div
              class="mantine-Group-root mantine-3suwep"
            >
              <div
                class="mantine-Group-root mantine-1u7f61y"
              >
                <div
                  class="mantine-SimpleGrid-root mantine-Group-child mantine-1qxm3qj"
                />
              </div>
              <div
                class="mantine-Group-root mantine-Group-child mantine-2v0bxn"
              >
                <div
                  class="flex"
                >
                  <button
                    class="absolute z-40 px-6 py-3 text-xl font-semibold text-white rounded-lg shadow-inner bg-ice-6 hover:mt-1 active:mt-2"
                  >
                    Ready!
                  </button>
                  <div
                    class="px-6 py-3 mt-2 text-xl font-semibold rounded-lg text-ice-7 bg-ice-7"
                  >
                    Ready!
                  </div>
                </div>
                <p
                  class="mantine-gnzaph mantine-Group-child text-center w-full text-red-500 hidden"
                >
                  Please choose a valid color
                </p>
              </div>
            </div>
          </div>
        </div>
        <div
          class="flex flex-col items-center justify-around w-3/5 xl:flex-row gap-y-2"
        >
          <div
            class="flex"
          >
            <button
              class="absolute z-40 px-6 py-3 text-xl font-semibold text-white rounded-lg shadow-inner bg-ice-6 hover:mt-1 active:mt-2"
            >
              Leave Room
            </button>
            <div
              class="px-6 py-3 mt-2 text-xl font-semibold rounded-lg text-ice-7 bg-ice-7"
            >
              Leave Room
            </div>
          </div>
          <div
            class="flex"
          >
            <button
              class="absolute z-40 px-6 py-3 text-xl font-semibold text-white rounded-lg shadow-inner bg-ice-6 hover:mt-1 active:mt-2"
            >
              Game Config
            </button>
            <div
              class="px-6 py-3 mt-2 text-xl font-semibold rounded-lg text-ice-7 bg-ice-7"
            >
              Game Config
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
